### DISCUSSION TITLE
Live Preview

### DISCUSSION THREAD

**Discussion by jacobsfletch** [payloadcms/payload#3061] [Roadmap]
We need to build a Live Preview feature into Payload so that editors can view updates on their front-end in real-time as they edit fields, without needing to save the document or navigate away from the admin dashboard. This will render a new "Preview" component directly alongside the document which will be used to render a website with live data on-the-fly.

How it will work:

Your website will be served within an iframe directly in the admin panel. As you make changes to your document, those changes are broadcast to the iframe via a `postMessage` event. Your website will then receive the event and re-render itself with the new data. Exactly how this is achieved will differ across front-end frameworks, but Payload will provide basic utility functions for the most popular ones, starting with Next.js.

What needs to be done:

The feature itself is relatively simple, but adjusting the admin UI to create extra space is not. Since the preview iframe will need to take up quite a lot of visual real estate, there is not enough room for _everything_ that our admin UI requires. So to mitigate this, we may need to drastically rearrange the UI to maximize horizontal space. Simply collapsing the sidebar will not be enough. Instead, the sidebar will likely be replaced with a header nav (pending design exploration).

Other requirements:

- Migrate sidebar to header (described above)
- Make sure live preview works with any frontend
- Include typical preview controls like dimensions, zoom, auto-update, etc
- Allow users to hide the fields so they can view their preview at full width
- Establish a pattern for multiple extensible and swappable “views”
- Make the UI **_NICE_**—use a border-less iframe, consider white-on-white, etc

Huge shoutout to @jpeiniger, @dkirchhof, @DennisSnijder, and their whole team for kickstarting this work with their [visual editor plugin](https://github.com/pemedia/payload-visual-editor). It's received some great reception and will be looked to for inspiration as we integrate this feature into core.

Related: https://github.com/payloadcms/payload/discussions/997

Update from @jacobsfletch 8/23/2023

We've begun redesigning the admin UI to accommodate Live Preview as well as the preview interface itself. There's a separate discussion around that (including images 🤩) which can be found here: https://github.com/payloadcms/payload/discussions/3217. Feel free to jump into the conversation there as well.

Update from @jacobsfletch 9/21/2023

[Admin 2.0](https://github.com/payloadcms/payload/discussions/3217) is done 🎉 
and there's a POC of Live Preview demonstrated here: https://github.com/payloadcms/payload/discussions/3061#discussioncomment-7074392

![live-preview-resized](https://github.com/payloadcms/payload/assets/65888/d6d90e37-80e3-4f68-8d51-0902c4b3819a)

**Comment by jpeiniger** [comment]
Looking forward to this becoming a core feature ❤️

**Comment by paulpopus** [comment]
Can we make this preview window popout-able?

Use case: On a laptop, you have a very limited amount of screen space and one of the most common usability features is alt-tabbing or virtual desktops so you can quickly switch between windows; the other one I use a lot is picture in picture.

Technical solutions:

1. Eventually, PIP support https://developer.chrome.com/docs/web-platform/document-picture-in-picture/
2. Until then, a button to open in a new tab/window? I can resize and pin that myself

**Reply by jacobsfletch** [reply]
@paulpopus this is 🔥

**Reply by matthijs166** [reply]
Very cool! I think it needs to be by default in the same window

**Reply by dkirchhof** [reply]
The extension of pip sounds great :)

**Comment by dkirchhof** [comment]
Just a quick test :upside_down_face: 

[visual-editor.webm](https://github.com/payloadcms/payload/assets/2778643/d95e915a-baf9-4a82-bc9f-4b0b0e716766)

**Comment by matthijs166** [comment]
Yeah UI is indeed the big question. I think we need to draw some inspiration from some other visual editors like: Webflow, Elementor, Wix etc..

It would also incredibly beneficial to send back the data within the live preview, allowing developers to create their own editing tools on top of it, for instance:
1. A title field defined in Payload as "Example" would be sent to the live preview, rendering the title as "Example" 
2. The developer can then wrap the title inside an edit component, allowing the admin to click on the title and modify it to say "Example Test" 
3. The updated title is then sent back (by the edit component) from the live preview to Payload, effectively saving the edit. 

By following this approach, developers can seamlessly integrate the live preview functionality into their applications, empowering users to make real-time changes without having to switch between different editing modes. This enhances the overall user experience and simplifies the editing process.

There is also the option to support/create the necessary edit components, but that's framework dependent, better to be left to the community.

Other thought is that we draw inspiration from https://vercel.com/docs/workflow-collaboration/visual-editing and use source maps to map the data and extend it with an overlay.

**Reply by jacobsfletch** [reply]
> Other thought is that we draw inspiration from https://vercel.com/docs/workflow-collaboration/visual-editing and use source maps to map the data and extend it with an overlay.

Hey @matthijs166 check this out 😉 https://payloadcms.com/docs/integrations/vercel-visual-editing

Whether its with Vercel or not, this is 100% the end-game. We will absolutely be building tools so we can send data both ways. It's much more tricky to go front-end->Payload though in a nice and elegant way that doesn't require serious overhaul of your components, i.e. Rich Text.

**Reply by dkirchhof** [reply]
Do you plan to allow the user to edit everything or just text / richt text fields?
Just remembered the pain in gutenberg where we have to define all components for the frontend and backend twice, because the styling doesn't match 100% etc.

I can imagine that editing plain text fields (for example by using the contenteditable attribute) would be fine.

For the preview window, I implemented the first messages, which will be send from the frontend to the cms.
This channel could be used to send changes from the iframe / window / pip back to the cms.
I have some ideas in mind, how it could work, but I think you are further than I think, right?

**Reply by jacobsfletch** [reply]
> Just remembered the pain in gutenberg where we have to define all components for the frontend and backend twice, because the styling doesn't match 100% etc.

Yea that's the tricky part here @dkirchhof, there needs to be as little manual effort as possible. This is why Vercel Visual Editing works well, they render hidden, encoded values into the front-end so that there is little to no integration needed. When these values are detected on the front-end, they overlay their own proprietary editor overtop so you can easily navigate back and forth between the CMS and the front-end. BUT this is not true _visual editing_, because as of now nothing is marked `contenteditable` (this will likely change in the future, if I had to guess). Payload will likely provide it's own tools for this, with text being relatively simply but rich text being extremely complex.

**Reply by franknoel** [reply]
@jacobsfletch Are you planning to release Vercel Visual Editing integration for non-enterprise customers?

**Comment by dkirchhof** [comment]
And another test:
Will playing around with a solution with less effort for the frontend implementation (like in vercels visual editing feature).

https://github.com/payloadcms/payload/assets/2778643/d8be206e-825f-4847-a97c-506ff367c5f6

**Reply by jpeiniger** [reply]
love it!

**Reply by dkirchhof** [reply]
And the other way around (select input => scroll to element in preview and highlight it):

https://github.com/payloadcms/payload/assets/2778643/9a60dcb8-7aa4-4070-9f9a-d6c05e801b9b

**Reply by coxmi** [reply]
Love it!

It'd be interesting to know how Vercel's visual editor deals with more complicated data types than string, or translated/processed values. 
I'd assume this will always need some form of manual data wrangling, since a date field could be "X minutes ago", ISO8601, a variety of international formats, and translated/processed values will be unpredictable.

**Comment by denolfe** [comment]
We'll be discussing Live Preview and Admin UI improvements today in Discord. https://twitter.com/payloadcms/status/1694047842560573569

![live-preview-resized](https://github.com/payloadcms/payload/assets/65888/25ec57dc-b3ec-43f5-adfe-96899aff53cd)

**Comment by jacobsfletch** [comment]
[Admin 2.0](https://github.com/payloadcms/payload/discussions/3217) has been completed! This has unlocked some major progress to Live Preview. Not quite ready to open a PR for this yet, but you can follow along [this branch](https://github.com/payloadcms/payload/tree/2_0/live-preview) for updates. Here's a sneak peak 😏:

https://github.com/payloadcms/payload/assets/15735305/39f105f0-cfc2-45e8-9679-fc4be5ed3251

**Reply by dkirchhof** [reply]
When I implemented the windowed preview in our plugin, I added a post message "channel" which is used to send message from the frontend back to payload. This is used to notify the CMS that the frontend is ready to receive updates. Payload will send immediately the current document afterwards.

This is just to prevent the "loading..." message before any changes made in the editor.

**Reply by jacobsfletch** [reply]
@dkirchhof yea 👍 that is exactly what the `useLivePreview` hook is doing currently here: https://github.com/payloadcms/payload/pull/3382/files#diff-b32ba70226aa90f43eda014925ed6d5bf5ab42bb54567f6cf6b5cef583b046f7R33

That loading message is actually related to the popup window not being ready before the initial post message is broadcast to it here: https://github.com/payloadcms/payload/pull/3382/files. I just wanted to share this before getting in the weeds with a fix for this.

Also, I did just draft a PR to make it easier to follow along: https://github.com/payloadcms/payload/pull/3382

**Comment by jacobsfletch** [comment]
Quick update here, Live Preview has been scaffolded out at a high level: https://github.com/payloadcms/payload/pull/3382. We're mainly through UI at this point (still a lot of refinements to be made here though). But we also have a lot of the patterns in place to begin finishing the API portion of this work, i.e. the `useLivePreview` hook. It's going to be a big week to get this finished out ahead of 2.0. I continue to post updates as I have them!

**Reply by jpeiniger** [reply]
That's awesome @jacobsfletch. Looking forward to it, as the API portion at the moment is a tricky part in the current plugin. E.g. getting relations inside of richtext and stuff like this. 😅

**Reply by franknoel** [reply]
Incredible work @jacobsfletch 👏 Can't wait to use it! One question that I always had about this feature is if it will work in a multiple-tenant project with different domains.

**Reply by jacobsfletch** [reply]
@franknoel hmmm this seems possible, we would just need to swap out the iframe src depending on the value the document's tenant. I can dive deeper into this some time, good call 👍

**Comment by jacobsfletch** [comment]
Big update here prior to the weekend: the data side of things is mostly complete! Andddd its written in a way that is _very_ composable. This is the part that I'm most excited about. This will enable us to support _any_ front-end framework, not just React. Essentially, we've abstracted the core functions out of the `useLivePreview` hook (like `subscribe` and `unsubscribe`) so that they can be integrated anywhere (like Vue or Svelte...or plain JS). I will be writing documentation around all of this, but you will be able to install a standalone dependency into your app that is per framework, starting with `@payloadcms/live-preview-react`.  Here's a recap of the work that went into this:

- [Builds `fieldSchemaToJSON` utility](https://github.com/payloadcms/payload/commit/f989e02a8558b1b85ec3e11ac62e11f76f4fc35c)
- [Handles live preview data](https://github.com/payloadcms/payload/commit/0ac36069bdd22eb7c00f55e414f043cd6eed383c)
- [Builds live preview JavaScript and React SDKs](https://github.com/payloadcms/payload/commit/b7d983f30a97be1cbd3c95edfb9848337adfb7c5)
- [Builds a more complex live preview test app](https://github.com/payloadcms/payload/commit/5f4c38ce21f8842404af5ff74b6b878a2e310643)
- [Supports live preview `url` as function](https://github.com/payloadcms/payload/commit/5bc81031d12832236a736255b329155199ec534b)
- And a few other misc fixes in between

This might also be the last update I make to this thread. This feature will be rolling out in beta **_next week!_** 😱 🚀

**Reply by jpeiniger** [reply]
Great work! Looking forward to playing around on Monday. 💪

**Reply by tiagordebarros** [reply]
Congratulations! Fantastic work!🚀

**Comment by robinscholz** [comment]
@jacobsfletch Is it correct to assume, that there is currently no way to enable live preview for data coming from custom endpoints? It seems like the feature is tightly coupled to the internal API, right?
